src/parameterizations/lateral/MOM_Zanna_Bolton.F90

```diff
@@ -173,726 +193,802 @@ subroutine ZB_2020_init(Time, GV, US, param_file, diag, CS, use_ZB2020)
+        intent(out) :: fy           !< Meridional acceleration due to convergence
+                                    !! of along-coordinate stress tensor [L T-2 ~> m s-2]
+  real, dimension(SZI_(G),SZJ_(G)),           &
+        intent(in) :: dx2h          !< dx^2 at h points [L2 ~> m2]
+  real, dimension(SZI_(G),SZJ_(G)),           &
+        intent(in) :: dy2h          !< dy^2 at h points [L2 ~> m2]
+  real, dimension(SZIB_(G),SZJB_(G)),         &
+        intent(in) :: dx2q          !< dx^2 at q points [L2 ~> m2]
+  real, dimension(SZIB_(G),SZJB_(G)),         &
+        intent(in) :: dy2q          !< dy^2 at q points [L2 ~> m2]
+
+  ! Local variables
+  real, dimension(SZI_(G),SZJ_(G)) :: &
+        Mxx, & ! Subgrid stress Txx multiplied by thickness and dy^2 [H L4 T-2 ~> m5 s-2]
+        Myy    ! Subgrid stress Tyy multiplied by thickness and dx^2 [H L4 T-2 ~> m5 s-2]
 
-    if (CS%id_S_12>0) then
-      do J=js-1,Jeq ; do I=is-1,Ieq
-        S_12_3d(I,J,k) = S_12(I,J)
-      enddo; enddo
-    endif
+  real, dimension(SZIB_(G),SZJB_(G)) :: &
+        Mxy    ! Subgrid stress Txy multiplied by thickness [H L2 T-2 ~> m3 s-2]
 
-    ! Weight with interface height (Line 1478 of MOM_hor_visc.F90)
-    ! Note that reduction is removed
-    do J=Jsq,Jeq+1 ; do i=Isq,Ieq+1
-      S_11(i,j) = S_11(i,j) * h(i,j,k)
-      S_22(i,j) = S_22(i,j) * h(i,j,k)
-    enddo ; enddo
+  real :: h_u ! Thickness interpolated to u points [H ~> m or kg m-2].
+  real :: h_v ! Thickness interpolated to v points [H ~> m or kg m-2].
+
+  real :: h_neglect    ! Thickness so small it can be lost in
+                       ! roundoff and so neglected [H ~> m or kg m-2]
+
+  integer :: is, ie, js, je, Isq, Ieq, Jsq, Jeq, nz
+  integer :: i, j, k
+
+  call cpu_clock_begin(CS%id_clock_divergence)
 
-    ! Free slip (Line 1487 of MOM_hor_visc.F90)
+  is  = G%isc  ; ie  = G%iec  ; js  = G%jsc  ; je  = G%jec ; nz = GV%ke
+  Isq = G%IscB ; Ieq = G%IecB ; Jsq = G%JscB ; Jeq = G%JecB
+
+  h_neglect  = GV%H_subroundoff
+
+  do k=1,nz
     do J=js-1,Jeq ; do I=is-1,Ieq
-      S_12(I,J) = S_12(I,J) * (hq(I,J) * G%mask2dBu(I,J))
+      if (CS%Klower_R_diss > 0) then
+        Mxy(I,J) = (CS%Txy(I,J,k) *                                         &
+                    (0.25 * ( (CS%c_diss(i,j  ,k) + CS%c_diss(i+1,j+1,k))   &
+                            + (CS%c_diss(i,j+1,k) + CS%c_diss(i+1,j  ,k)))  &
+                    )                                                       &
+                   ) * CS%hq(I,J,k)
+      else
+        Mxy(I,J) = CS%Txy(I,J,k) * CS%hq(I,J,k)
+      endif
+    enddo ; enddo
+
+    do j=js-1,je+1 ; do i=is-1,ie+1
+      if (CS%Klower_R_diss > 0) then
+        Mxx(i,j) = ((CS%Txx(i,j,k) * CS%c_diss(i,j,k)) * h(i,j,k)) * dy2h(i,j)
+        Myy(i,j) = ((CS%Tyy(i,j,k) * CS%c_diss(i,j,k)) * h(i,j,k)) * dx2h(i,j)
+      else
+        Mxx(i,j) = ((CS%Txx(i,j,k)) * h(i,j,k)) * dy2h(i,j)
+        Myy(i,j) = ((CS%Tyy(i,j,k)) * h(i,j,k)) * dx2h(i,j)
+      endif
```

Comment:
You may want to move the if-blocks outside of the loop for performance.

Reply:
As far as I know, these ifs, which evaluate the same at every loop iteration, almost do not affect the performance because of the branch prediction. They may be a bit slower for very small loops. 

I played a bit with these (and similar) if statements while preparing the PR and did not find substantial dependence of performance on the order of if and do statements.

BTW, all these loops are vectorized.

Reply:
I think the result can depend on both compiler and the hardware (e.g. Intel masking instructions are about 6x faster than AMD).  But if you don't believe it's an issue then no worries.

Reply:
Also FWIW, branch prediction is also going to vary across hardware.  For example, an Intel or AMD chip is probably going to have better prediction and a GPU or an A64FX core.

Reply:
I just tested this in GNU 13, and inverting does appear to make a difference.

The original loop:
```
.L554:
    vmovsd  (%r9,%rax,8), %xmm0
    vmulsd  (%rbx,%rax,8), %xmm0, %xmm2
    vmulsd  (%r10,%rax,8), %xmm0, %xmm0
    vmovsd  (%r8,%rax,8), %xmm6
    vmulsd  %xmm6, %xmm2, %xmm2
    vmulsd  %xmm6, %xmm0, %xmm0
    vmulsd  (%rdx,%rax,8), %xmm2, %xmm2
    vmulsd  (%r11,%rax,8), %xmm0, %xmm0
    vmovsd  %xmm2, (%rsi,%rax,8)
    vmovsd  %xmm0, (%rcx,%rax,8)
    addq    $1, %rax
    leal    -1(%rax), %r14d
    cmpl    %r14d, %edi
    jge .L554
```
Although it's using AVX instructions, they are not parallel (mulsd vs mulpd)

Inverting the if/do blocks gave me this:
```
.L550:
    vmovupd (%r11,%rax), %ymm7
    vmulpd  (%r14,%rax), %ymm7, %ymm0
    vmulpd  (%r12,%rax), %ymm0, %ymm0
    vmovupd (%r10,%rax), %ymm7
    vmulpd  (%rbx,%rax), %ymm0, %ymm0
    vmovupd %ymm0, (%rdx,%rax)
    vmulpd  (%r14,%rax), %ymm7, %ymm0
    vmulpd  (%r12,%rax), %ymm0, %ymm0
    vmulpd  (%r9,%rax), %ymm0, %ymm0
    vmovupd %ymm0, (%rcx,%rax)
    addq    $32, %rax
    cmpq    %rax, %rsi
    jne .L550
```
This loop is using the parallel instructions.

It is possible that Intel is a bit smarter and managed to vectorize the loop, but if that is the case then it would seem to be a small amount of work to get better performance across compilers.

So I would still recommend inverting the loops, but it's up to you.

Reply:
Thanks for the explanations!

I updated these loops. Also I removed ifs from the [loop](https://github.com/Pperezhogin/MOM6/blob/PR_ZB_2020_NW2_and_acceleration_combined/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L501). I have another one place with ifs in [loop](https://github.com/Pperezhogin/MOM6/blob/PR_ZB_2020_NW2_and_acceleration_combined/src/parameterizations/lateral/MOM_Zanna_Bolton.F90#L586). In this case, I cannot remove ifs without significant duplication of the code (a separate loop for every combination of if conditions) or storing intermediate computations into 2D slices (which is likely less memory efficient).